### DISCUSSION TITLE
Support property-based testing

### DISCUSSION THREAD

**Discussion by reta** [instancio/instancio#368] [Ideas]
### Feature description

What is property-based testing? According to [Wikipedia](https://en.wikipedia.org/wiki/Software_testing#Property_testing):

> Property based testing is a testing technique where, instead of asserting that specific inputs produce specific expected outputs, the practitioner randomly generates many inputs, runs the program on all of them, and asserts the truth of some "property" that should be true for every pair of input and output.

### Why Instancio?
[Instancio](https://github.com/instancio/instancio) provides extensive means to create and populates objects, randomly. This is one of the foundation property-based testing relies on: randomly generate many inputs. The missing part - support the semantics of property-based testing, which includes:
 - property
 - conditional generation (declarative or programmatic)
 - assertions
 - arbitraries 
 - ...

Alternatively to own implementation, [Instancio](https://github.com/instancio/instancio) could integrate with existing PBT libraries to provide the generation capabilities. 

### Other libraries (Java)
- https://github.com/jlink/jqwik
- https://github.com/pholser/junit-quickcheck

**Comment by reta** [comment]
@armandino opening up your idea for discussion :-)

**Reply by armandino** [reply]
@reta thanks for kicking it off! Let's move this one to an internal discussion. I created one earlier but forgot to tag you :)

**Reply by reta** [reply]
> @reta thanks for kicking it off! Let's move this one to an internal discussion. I created one earlier but forgot to tag you :)

Oh sorry, I haven't see it, but now I got the notification, thank you!
